src/mapreduce.jl

```diff
@@ -8,16 +8,21 @@
 # `map(f, as::Union{<:StaticArray,AbstractArray}...)` which included at least one `StaticArray`
 # this is not the case on 0.7 and we instead hope to find a StaticArray in the first two arguments.
 @inline function map(f, a1::StaticArray, as::AbstractArray...)
-    _map(f, same_size(a1, as...), a1, as...)
+    _map(f, a1, as...)
 end
 @inline function map(f, a1::AbstractArray, a2::StaticArray, as::AbstractArray...)
-    _map(f, same_size(a1, a2, as...), a1, a2, as...)
+    _map(f, a1, a2, as...)
 end
 @inline function map(f, a1::StaticArray, a2::StaticArray, as::AbstractArray...)
-    _map(f, same_size(a1, a2, as...), a1, a2, as...)
+    _map(f, a1, a2, as...)
 end
 
-@generated function _map(f, ::Size{S}, a::AbstractArray...) where {S}
+@generated function _map(f, a::AbstractArray...)
+    i = findfirst(ai -> ai <: StaticArray, a)
+    if i === nothing
+        return :(throw(ArgumentError("No StaticArray found in argument list")))
+    end
+    S = Size(a[i])
```

Comment:
Is it ok to call `Size` inside the generator here? User defined wrapper types might add methods to this trait.

Though it looks like the current version of `same_size` wouldn't work for those anyway because it uses `_first_static`. Hmm.

Reply:
I'm afraid I don't quite understand what the issue might be here. Could you elaborate?

Reply:
The issue is that generators run in the world as it existed when they were defined, so you can't call functions inside the generator which might have new methods added:

```julia
foo(x) = "default"
@generated function bar(x)
    foo(x) # Function foo is frozen in world where bar is defined
end

foo(::Type{Int}) = "override"
@generated function bar2(x)
    foo(x) # Otherwise identical `bar2` sees both foo methods.
end
```

Thence:
```julia
julia> bar(1)
"default"

julia> bar2(1)
"override"
```

It's quite a restriction :-/ This is why we need the pattern of constructing the traits outside the generators and passing them in as types.

Reply:
Ah i see where you getting at. But we (currently) only dispatch to `_map` if at least one of the first two arguments is a `StaticArray`, and `a[i]` will be the first of those. Further, for a `StaticArray` the implementation from https://github.com/JuliaArrays/StaticArrays.jl/blob/bb82cd5b78bb0945565652ebe5b23116e33eb3ab/src/traits.jl#L88 should always be correct. If a subtype overloads `Size` to give a different size than what it declares in its supertype, that's just ... wrong.

But if we in the future want to dispatch wrapped `StaticArray`s, where the wrapper resides in some other package, to `_map`, that would become a problem indeed. At the moment, I think the only option for that dispatch to happen would be for that other package to provide a `map` that invokes `StaticArrays._map`. Which raises the question: Do such packages exist and should we therefore keep the old `_map` signature for compatibility anyway?

Reply:
> But we (currently) only dispatch to _map if at least one of the first two arguments is a StaticArray, and `a[i]` will be the first of those

I think you're right that this is quite unlikely to cause problems with things as they exist right now. Maybe we should just go with it, with a comment clearly noting that calling this in the generator is only ok because of the particular restrictions on which `a[i]` we pick out.

> other package to provide a map that invokes `StaticArrays._map`

I don't know of any packages using `_map`. But if they do see fit to use these internal implementation details, I think the onus is on them to keep up to date with `StaticArrays` :-)